backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java

```diff
@@ -108,6 +114,68 @@ public void setTaxAmtInfo(final Properties ctx, final I_C_InvoiceLine il, final
 	@Override
 	public boolean setTax(final Properties ctx, final org.compiere.model.I_C_InvoiceLine il, final String trxName)
 	{
+		final IInOutDAO inoutDAO = Services.get(IInOutDAO.class);
+		final IInvoiceDAO invoiceDAO = Services.get(IInvoiceDAO.class);
+		final IWarehouseBL warehouseBL = Services.get(IWarehouseBL.class);
+
+		final InvoiceId invoiceId = InvoiceId.ofRepoId(il.getC_Invoice_ID());
+		final I_C_Invoice invoice = invoiceDAO.getByIdInTrx(invoiceId);
+
+		if (il.getM_InOutLine_ID() <= 0)
+		{
+			logger.debug(il + "has M_InOutLine_ID=" + il.getM_InOutLine_ID() + ": returning");
+			return false;
+		}
+
+		final InOutLineId inoutLineId = InOutLineId.ofRepoId(il.getM_InOutLine_ID());
+		final I_M_InOutLine inoutLineRecord = inoutDAO.getLineById(inoutLineId);
+
+		final I_M_InOut io = inoutDAO.getById(InOutId.ofRepoId(inoutLineRecord.getM_InOut_ID()));
+
+		final OrgId inOutOrgId = OrgId.ofRepoId(io.getAD_Org_ID());
+
+		final Timestamp shipDate = io.getMovementDate();
+
+		final BPartnerLocationId partnerLocationId = BPartnerLocationId.ofRepoId(io.getC_BPartner_ID(), io.getC_BPartner_Location_ID());
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoId(io.getM_Warehouse_ID());
+
+		final boolean isSOTrx = io.isSOTrx();
+
+		final CountryId countryFromId;
+		final WarehouseId invoiceWarehouseId = WarehouseId.ofRepoIdOrNull(invoice.getM_Warehouse_ID());
+
+		if (invoiceWarehouseId != null)
+		{
+			countryFromId = warehouseBL.getCountryId(invoiceWarehouseId);
+		}
+		else
+		{
+			countryFromId = warehouseBL.getCountryId(warehouseId);
+		}
+
+		return setTaxForInvoiceLine(ctx, il, inOutOrgId, shipDate, countryFromId, partnerLocationId, isSOTrx);
+	}
+
+	@Override
+	public boolean setTaxForInvoiceLine(final Properties ctx,
+			final org.compiere.model.I_C_InvoiceLine il,
+			final OrgId orgId,
+			final Timestamp taxDate,
+			final CountryId countryFromId,
+			final BPartnerLocationId partnerLocationId,
```

Comment:
rename `partnerLocationId` parameter to `billLocationId` or similar (to underline that we are talking about BILL BPartner and location)

Reply:
... or is this the shipToLocation? can u pls clarify?

Reply:
Hi, @teosarca 
* I checked and saw that before, we took the shipment bp and location whensetting the tax from ```de.metas.invoice.service.impl.InvoiceLineBL.setTax(I_C_InvoiceLine, String)``` and the invoice bp and location when setting it from ```org.compiere.model.MInvoiceLine.setTax()```
* Note that in the latter one the inout is not even searched for.
* Which partner and location should be considered when computing the tax? How is this legally done in the real world? 
* Similar: Which date should be used on tax computation? The shipment date or the invoice date?

Reply:
I renamed the method to make it obvious the data is based on the inoutLine linked to the invoice. If there is no inoutline, the tax is not set.